src/shell/shell_misc.cpp

```diff
@@ -20,479 +20,421 @@
-			break;
-		case'\t':
-			{
-				if (l_completion.size()) {
-					++it_completion;
-					if (it_completion == l_completion.end())
-						it_completion = l_completion.begin();
-				} else {
-					// build new completion list
-					// Lines starting with CD will only get
-					// directories in the list
-					bool dir_only = (strncasecmp(line,"CD ",3)==0);
-
-					// get completion mask
-					char *p_completion_start = strrchr(line, ' ');
-
-					if (p_completion_start) {
-						p_completion_start ++;
-						completion_index = (uint16_t)(str_len - strlen(p_completion_start));
-					} else {
-						p_completion_start = line;
-						completion_index = 0;
-					}
-
-					char *path;
-					if ((path = strrchr(line+completion_index,'\\'))) completion_index = (uint16_t)(path-line+1);
-					if ((path = strrchr(line+completion_index,'/'))) completion_index = (uint16_t)(path-line+1);
-
-					// build the completion list
-					char mask[DOS_PATHLENGTH] = {0};
-					if (p_completion_start) {
-						if (strlen(p_completion_start) + 3 >= DOS_PATHLENGTH) {
-							//Beep;
-							break;
-						}
-						safe_strcpy(mask, p_completion_start);
-						char* dot_pos=strrchr(mask,'.');
-						char* bs_pos=strrchr(mask,'\\');
-						char* fs_pos=strrchr(mask,'/');
-						char* cl_pos=strrchr(mask,':');
-						// not perfect when line already contains wildcards, but works
-						if ((dot_pos-bs_pos>0) && (dot_pos-fs_pos>0) && (dot_pos-cl_pos>0))
-							strncat(mask, "*",DOS_PATHLENGTH - 1);
-						else strncat(mask, "*.*",DOS_PATHLENGTH - 1);
-					} else {
-					        safe_strcpy(mask, "*.*");
-				        }
-
-					RealPt save_dta=dos.dta();
-					dos.dta(dos.tables.tempdta);
-
-					bool res = DOS_FindFirst(mask, 0xffff & ~DOS_ATTR_VOLUME);
-					if (!res) {
-						dos.dta(save_dta);
-						break;	// TODO: beep
-					}
-
-					DOS_DTA dta(dos.dta());
-					char name[DOS_NAMELENGTH_ASCII];uint32_t sz;uint16_t date;uint16_t time;uint8_t att;
-
-					std::list<std::string> executable;
-					while (res) {
-						dta.GetResult(name,sz,date,time,att);
-						// add result to completion list
-
-						if (strcmp(name, ".") && strcmp(name, "..")) {
-							if (dir_only) { //Handle the dir only case different (line starts with cd)
-								if(att & DOS_ATTR_DIRECTORY) l_completion.push_back(name);
-							} else {
-							        if (is_executable_filename(name))
-								        // Prepend executables to a separate list
-								        // and place that list ahead of normal files.
-									executable.push_front(name);
-								else
-									l_completion.push_back(name);
-							}
-						}
-						res=DOS_FindNext();
-					}
-					/* Add executable list to front of completion list. */
-					std::copy(executable.begin(),executable.end(),std::front_inserter(l_completion));
-					it_completion = l_completion.begin();
-					dos.dta(save_dta);
+		}
+
+		case '\t': {
+			if (!completion.empty()) {
+				completion_index++;
+				if (completion_index == completion.size()) {
```

Comment:
Do we expect that the incremented `completion_index` will never exceed completion.size()?

If it's a valid risk, maybe: 
`if (completion_index >= completion.size())`

If we can be more certain, suggest asserting that prior to the if-check.